### PR TITLE
new resource "azurerm_key_vault_key_encrypt" and data source "azurerm_key_vault_key_decrypt"

### LINKED ISSUE #7856: Support for decrypt secrets from data encrypted with Azure keyvault key
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS provider has this data source which decrypts cipherblob and then stores it in plain text within secrets manager. Need similar functionality for azurerm, which will decrypt the cipherblob (using keyvault key) and then store it in plain text in keyvault secrets.

https://www.terraform.io/docs/providers/aws/d/kms_secrets.html

### PR DESCRIPTION
**Pull request by njuCZ**
fix #7856

resource "azurerm_key_vault_key_encrypt" is similar with "aws_kms_ciphertext"
data source "azurerm_key_vault_key_decrypt" is similar with "aws_kms_secret"

to keep a stable encrypted value, make "azurerm_key_vault_key_encrypt" be a resource but not a data source.
there is no need to support "import" operation for resource "azurerm_key_vault_key_encrypt"

image

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
dependent on https://github.com/terraform-providers/terraform-provider-azurerm/issues/7856#issuecomment-711619498

**Comment by saranshlamba** [comment]
@tombuildsstuff Good morning. Any updates on this and when it will be merged?

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package keyvault
```
Comment (IanMoroney) [inline azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:1]: If the intention is to encrypt secrets, might i suggest that the resource instead be called:
`azurerm_key_vault_secret_encrypt`

File: azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
+package keyvault
```
Comment (IanMoroney) [inline azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go:1]: If the intention is to decrypt secrets, might i suggest that the resource instead be called:
`azurerm_key_vault_secret_decrypt`

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
```
Comment (IanMoroney) [inline azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:2]: I can see there being a value in having a `resource` block for encrypting secrets with a static encrypted blob, and a `data` block for encrypting secrets with a value that changes on every apply, but I only see an `encrypt` resource. 

Could you create an equivalent `encrypt` data block, as if we're modelling it after aws_kms_ciphertext, we should have the ability for users to choose whether to change the encrypted blob frequently or infrequently.

File: azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
+			"key_vault_key_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: azure.ValidateKeyVaultChildId,
+			},
+
+			"payload": {
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go:31]: what format are we expecting here? base64?
```suggestion
			"encrypted_base64_data": {
```

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(keyvault.RSA15),
+					string(keyvault.RSAOAEP),
+					string(keyvault.RSAOAEP256),
+				}, false),
+			},
+
+			"cipher_text": {
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:60]: should this also then be
```suggestion
			"encrypted_data_in_base64": {
```

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			// avoid test fail, add timeout config for read and delete
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:34]: does it make sense to add a name property so we can identify this and use it in the ID?

File: azurerm/internal/services/keyvault/tests/key_vault_key_decrypt_data_source_test.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+		},
+	})
+}
+
+func testAccDataSourceKeyVaultKeyDecrypt_basic(data acceptance.TestData, plaintext string) string {
+	t := testAccAzureRMKeyVaultKeyEncrypt_basic(data, plaintext)
+	return fmt.Sprintf(`
+%s
+
+data "azurerm_key_vault_key_decrypt" "test" {
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/tests/key_vault_key_decrypt_data_source_test.go:33]: could we use the encrypt resource and decrypt to check that the workflow works?

File: website/docs/d/key_vault_key_decrypt.html.markdown [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_key_vault_key_decrypt"
+description: |-
+  Decrypt data encrypted with the Key Vault Key.
+
+---
+
+# Data Source: azurerm_key_vault_key_decrypt
+
+Use this data source to decrypt data encrypted with the Key Vault Key.
```
Comment (katbyte) [inline website/docs/d/key_vault_key_decrypt.html.markdown:12]: grammer
```suggestion
Use this data source to decrypt data encrypted with a Key Vault Key.
```

File: website/docs/r/key_vault_key_encrypt.html.markdown [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_key_vault_key_encrypt"
+description: |-
+  Encrypt data with the Key Vault Key.
+
+---
+
+# azurerm_key_vault_key_encrypt
+
+Use this resource to encrypt data with the Key Vault Key.
```
Comment (katbyte) [inline website/docs/r/key_vault_key_encrypt.html.markdown:12]: grammer

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @njuCZ - left some comments inline

File: azurerm/internal/services/keyvault/tests/key_vault_key_decrypt_data_source_test.go (hunk 7 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/tests/key_vault_key_decrypt_data_source_test.go:33]: yes, in this data source, we assign the `cipher_text` of `azurerm_key_vault_key_encrypt` to `payload`, and in the check functions, we checked whether the descrypt value is the inital input plaintext.
In this way, we checked the whole encrypt and decrypt workflow

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go (hunk 5 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:66]: the result format is `base64url`, which is slightly different from `base64`. please refer to RFC: https://tools.ietf.org/html/rfc4648#section-5. I have noted it in the doc. Should I still rename this property?

File: azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go (hunk 4 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go:32]: as below

**Comment by njuCZ** [comment]
Thanks @katbyte , I have rebased and refine this PR

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go (hunk 5 above)
Comment (katbyte) [reply azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:69]: yes lets rename it to match what is expected so people don't have to look at docs and can tell from the property name. as for url in there is this different then any other base64 property for azure?

File: azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+			"key_vault_key_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NestedItemId,
+			},
+
+			"payload": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go:35]: lets just use base64 without any url in the property name and then validate it?

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go (hunk 5 above)
Comment (katbyte) [reply azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:69]: lets go with `encrypted_data_in_base64` and then add a validation function to ensure its the right base64 format

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by njuCZ** [comment]
@katbyte Thanks for your opinion, I have renamed the field and added validation

**Comment by IanMoroney** [comment]
Any thoughts on my comments @njucz?

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go (hunk 1 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:1]: becasue key vault key and key vault secret are different resource, and the rest api is using key vault key to encrypt and decrypt data. So personally I think `azurerm_key_vault_secret_encrypt` maybe a little misleading?

File: azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go (hunk 2 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_key_decrypt_data_source.go:1]: as above

File: azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go (hunk 3 above)
Comment (njuCZ) [reply azurerm/internal/services/keyvault/key_vault_key_encrypt_resource.go:2]: is there any scenario that users want to change the encrypted data every time? In the `resource`, we could also change the encrypted data by changing the property `name`

**Comment by njuCZ** [comment]
@IanMoroney sorry for late reply. I thought this PR was blocked, but it seems things become better now. I have commented inline. Thanks for your review

**Comment by gro1m** [comment]
@njuCZ @IanMoroney What is the status here, actually?

**Comment by tombuildsstuff** [comment]
hi @njuCZ 

Apologies for the delayed re-review of this PR - taking a look through here whilst this is functional these both want to be data sources and since the value can be converted bi-directionally we can make these a singular data source for "encrypted value".

Since there's some larger changes to get this PR in, I hope you don't mind but I've opened #15873 which supersedes this PR and will allow this to ship via the `azurerm_key_vault_encrypted_value` data source, which as mentioned above allows a bi-directional encryption/decryption of a plain-text/encrypted value as necessary - as such whilst I'd like to thank you for this contribution I'm going to close this in favour of #15873

Thanks!

**Pull request closed without merging**